torch/fx/experimental/symbolic_shapes.py

```diff
@@ -811,8 +811,15 @@ class EqualityConstraint(Constraint):
     Given pairs of sources corresponding to pairs of dynamic dimensions that
     are specified equal, represent them in a union-find data structure so that
     we can efficiently check whether two such sources are transitively equal.
+
+    Also contains a list of derived equalities, each of which relates a source
+    to an expression over a root. The root can be a source, corresponding to some
+    input shape, or a phantom symbol corresponding to the root Dim of a derived
+    Dim describing some input shape. We track the names of such Dims as well.
```

Comment:
I'm not sure I understand this explanation of phantom. An example would go a long way here!

Reply:
A phantom symbol is a symbol that isn't directly the shape of an input tensor, but appears in a symbolic expression that describes such a shape. E.g., an input tensor might have shape `2*s0` where `s0` is not the shape of any input tensor. In that case `s0` is a phantom and doesn't have a source.

Reply:
One thing that concerns me about phantom symbols is that you can easily generate non-instantiatable sizevars. For example, let's say I declare a single tensor input with size `[s0 + s1]`, it is impossible to know what s0 and s1 are, because the phantoms are underconstrained. I wonder what you are doing about this.

Reply:
The input language for derived dims is deliberately very constrained. One variable, linear expressions of the form `a*x + b`. So this kind of situation cannot happen.

Reply:
I guess (1) people will probably eventually want to added other forms for derived dims, so if this is the invariant we will have to make sure we keep upholding this invariant, and (2) even with linear expressions, it is easy to give an unsolvable equation (since x must be integral), so I will looking out to see where we check for this.